### ISSUE TITLE
[Experiment] Changing Conv2D Kernel initialization

### ISSUE THREAD

**Issue by sethtroisi** [tensorflow/minigo#416]
I was (and continue to be) inspired by @lightvector's amazing work on [GoNN](https://github.com/lightvector/GoNN/). One hypothesis lightvector tested was to increase the weights near the center of Conv2D kernels. ([see writeup here](https://github.com/lightvector/GoNN/#redundant-convolution-weights)).

[I coded this up](https://github.com/tensorflow/minigo/pull/415/files) for Minigo and started testing it. From some past work (https://github.com/tensorflow/minigo/issues/350) and best ML practices I decided to test each training run multiple times.

### hyperparams*:
* 4 runs, 10 blocks, 128 filters, 10,240 steps (20M examples**) 
* **center_conv_kernel_weight**  of **0.5, 1.0, 1.2, 1.5, and 2.0**

## Initial results: Nothing but noise(?)
* Two runs with identical settings (but different random seeds) vary just as much different center_conv_kernel_weight
* L2 loss changes as expected so code seems to work but policy and value don't seem to converge faster.

![image](https://user-images.githubusercontent.com/10172976/45236349-78694600-b290-11e8-9bfd-f6d769ff92b9.png)

![image](https://user-images.githubusercontent.com/10172976/45236275-3dffa900-b290-11e8-957c-122bc6224fdc.png)

![image](https://user-images.githubusercontent.com/10172976/45236257-34764100-b290-11e8-8abe-e7d6e21d246f.png)

----

\* Code used to test
```
train() {
  BOARD_SIZE=19 python dual_net.py train \
       --use_tpu --tpu_name=sethtroisi$1 \
       --model_dir=$WORK_DIR/conv_weight/$2_$3_$4_run_$1 \
       --steps=10240 --iterations_per_loop=128 --summary_steps=256 \
       --trunk_layers=$2 --conv_width=$3 --conv_kernel_center_scale=$4 \
       $DATA_DIR/{250..259}.tfrecord.zz
}
tpu=$1
train $tpu 10 128 0.5
train $tpu 10 128 1.0
train $tpu 10 128 1.2
train $tpu 10 128 1.5
train $tpu 10 128 2
```

\*\* I used preprocessed golden chunks 250 to 259 from v9 which was a reasonable strong model. If I repeated this experiment I'd use a later golden chunks.

**Comment by lightvector** [comment]
Thanks for running this, interesting to see the negative result! 

I guess I'm not too surprised at this, since the effect was at most small when I tested it. I also didn't look separately at the effect of the initialization of the center weight larger versus increasing the learning rate on the center - due to limited computing power I did them together so it's possible that only one of them matters and not the other. 

But I also don't think I did a particularly large number of training trials with the new settings (I think just one or two independent initializations), so it could certainly have been just variation in those runs rather than a real effect.

Is it easy for you to try the version with the learning rate modification rather than the differing initialization? If you would find it useful, I could also revisit the idea myself to see if it holds up again after all the changes and size increases that have happened between then and now for my neural net.

**Comment by lightvector** [comment]
Since this post made me personally curious whether the results I found earlier would reproduce, especially now that I'm using a much larger neural net than before, 12 blocks instead of 5 blocks, I ran two new short training runs (only 40 million training samples). In both of these training runs:
   1. I removed the 1.5x learning rate factor I've been using on the center weight of each convolution for the residual blocks that I've been using ever since my earlier results.
   2. I removed an extra 1x1 center parameter (equivalent to a combined sqrt(2)x initialization plus a 2x learning rate on the center weight) that I had for the initial convolution of the neural net that I've also been using ever since those earlier results. 
   3. I did NOT remove the larger center initialization for these all of these convolutions that I've been using

So except for the detail about the initial convolution in # 2, these changes are *mostly* a test of the effect of the learning rate of the center of the convolution, rather than the initialization of the center.

I compared these two runs ("value37","value38") to two of my other earlier runs that I already had data logged for in which I did not make these removals, ("value33","value28") but are otherwise the same. 
<sub>As a minor detail, value28 also has a slight change to the head architecture that as far as I can tell has no detectable effect on anything, if you're paranoid this maybe affects the comparison, but I went ahead with it since I already had the data and it would save the cost and time of doing a third new short run.</sub>

Here is the validation policy loss plotted by millions of training samples, as well as a graph of the running difference of each run from the mean of all four. The red and yellow are the ones with the learning rate on the center loss removed.
![vplosscenterlr](https://user-images.githubusercontent.com/11942395/45398371-d049d900-b611-11e8-8996-9910f7c45eab.png)

So the difference is pretty slight, but seems to be there. One confounding factor that I haven't tried to tease apart is if maybe simply slightly increasing the learning rate for *everything* would produce a similar result (e.g. is my learning rate slightly too slow at this point given the current architecture?).

Here's a plot of the training value loss which shows a similar effect. (For me, the validation value loss is easily one to two orders of magnitude too noisy at this point in training, but the training value loss is non-noisy enough to look at).

![tvlosscenterlr](https://user-images.githubusercontent.com/11942395/45398932-2881da80-b614-11e8-9e52-4d1c5a9f4967.png)

**Comment by sethtroisi** [comment]
I suspect that you've just bumped up the global learning rate.

I'm repeated doing another twenty runs with LR {0.01, 0.012, 0.014, 0.0167, and 0.02}. I'm seeing a similar shift downwards (or compression in steps) of policy and value.

Some initial results
![image](https://user-images.githubusercontent.com/10172976/45467808-5af40c00-b6d6-11e8-934c-0638cc415be4.png)

![image](https://user-images.githubusercontent.com/10172976/45467860-95f63f80-b6d6-11e8-94c9-fd6c1287e284.png)


If you change only the LR of the center convolution it's probably only a ~5% increase in global LR so the effect will be magnified by my results. If you'd like I can repeat with different LR (or even a schedule of very fast, then slow).

**Comment by lightvector** [comment]
Yep, that's what the idea always was - that increasing the center LR would speed up training, but not necessarily make the final neural net stronger after "convergence".

I did before all this tune learning rate a bit to try to optimize training speed over the first hundred or so million training samples, but not in increments quite as fine as a few percent. Continuing the curiosity here I tried seeing the effect of still having the removing the center LR bonus removed but then also globally bumping learning rate by 5%, so far it looks like that doesn't produce quite the same the same slight improvement in loss, although it did shift the value loss down a little. Possibly the "equivalent" global LR bump is a bit larger than 5%, or possibly since the neural net does need to walk "farther" in parameter space along the "center weights of convolutions" axes to get to a good set of weights, increasing the learning rate there alone gets you a slight improvement in training speed while having less additional instability that you normally get as you turn up the learning rate too high.

If it does though, seems like the effect is very small, probably not worth pursuing much further. Maybe I'll run one more test with another different global LR to see if I can reproduce the effect purely with global LR, but beyond that it's probably more worthwhile to free up my GPUs for other things I still want to test.
